File: backend/src/models/Vehicle.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

const VEHICLE_TYPES = [
  'flatbed', // مسطحة
  'tipper', // قلابة
  'semi_trailer', // نصف مقطورة
  'car_carrier', // حاملة آليات
  'refrigerated', // مبردة
  'tanker', // صهريج
  'closed', // مغلقة
];

const vehicleSchema = new Schema(
  {
    carrierId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    plateNumber: { type: String, required: true, trim: true },
    brand: String,
    model: String,
    type: { type: String, enum: VEHICLE_TYPES, required: true },
    payloadCapacityKg: Number,
    yearOfManufacture: Number,
    photos: [{ url: String, publicId: String }],

    documents: [
      {
        type: { type: String, enum: ['registration', 'insurance', 'technical_inspection'] },
        url: String,
        publicId: String,
        expiresAt: Date,
        uploadedAt: { type: Date, default: Date.now },
      },
    ],

    status: {
      type: String,
      enum: ['available', 'on_mission', 'broken_down', 'maintenance'],
      default: 'available',
    },

    assignedDriverId: { type: Schema.Types.ObjectId, ref: 'User' },
  },
  { timestamps: true }
);

module.exports = mongoose.model('Vehicle', vehicleSchema);
module.exports.VEHICLE_TYPES = VEHICLE_TYPES;

File: mobile_app/lib/features/carrier/models/vehicle.dart
/// Mirrors backend Vehicle model (no dedicated core model exists yet, so this
/// lives under the carrier feature — only carrier screens use it).
class VehicleDoc {
  VehicleDoc({required this.type, this.expiresAt, this.url});

  final String type; // registration | insurance | technical_inspection
  final DateTime? expiresAt;
  final String? url;

  factory VehicleDoc.fromJson(Map<String, dynamic> json) => VehicleDoc(
        type: json['type'] ?? '',
        expiresAt: json['expiresAt'] != null ? DateTime.tryParse(json['expiresAt']) : null,
        url: json['url'],
      );
}

class Vehicle {
  Vehicle({
    required this.id,
    required this.plateNumber,
    this.brand,
    this.model,
    required this.type,
    this.payloadCapacityKg,
    this.yearOfManufacture,
    this.status = 'available',
    this.documents = const [],
  });

  final String id;
  final String plateNumber;
  final String? brand;
  final String? model;
  final String type;
  final double? payloadCapacityKg;
  final int? yearOfManufacture;
  final String status; // available | on_mission | broken_down | maintenance
  final List<VehicleDoc> documents;

  factory Vehicle.fromJson(Map<String, dynamic> json) => Vehicle(
        id: json['_id'] ?? json['id'] ?? '',
        plateNumber: json['plateNumber'] ?? '',
        brand: json['brand'],
        model: json['model'],
        type: json['type'] ?? 'flatbed',
        payloadCapacityKg: (json['payloadCapacityKg'] as num?)?.toDouble(),
        yearOfManufacture: json['yearOfManufacture'] is int
            ? json['yearOfManufacture']
            : int.tryParse('${json['yearOfManufacture'] ?? ''}'),
        status: json['status'] ?? 'available',
        documents: (json['documents'] as List? ?? []).map((d) => VehicleDoc.fromJson(d)).toList(),
      );

  /// Nearest document expiry across all uploaded docs, if any.
  DateTime? get nearestExpiry {
    final dates = documents.map((d) => d.expiresAt).whereType<DateTime>().toList();
    if (dates.isEmpty) return null;
    dates.sort();
    return dates.first;
  }
}

/// Mirrors backend Driver (a user with role=driver, created/owned by a carrier).
class CarrierDriver {
  CarrierDriver({
    required this.id,
    this.fullName,
    required this.phone,
    this.licenseNumber,
    this.licenseCategory,
    this.licenseExpiresAt,
    this.status = 'pending',
  });

  final String id;
  final String? fullName;
  final String phone;
  final String? licenseNumber;
  final String? licenseCategory;
  final DateTime? licenseExpiresAt;
  final String status;

  factory CarrierDriver.fromJson(Map<String, dynamic> json) => CarrierDriver(
        id: json['_id'] ?? json['id'] ?? '',
        fullName: json['fullName'],
        phone: json['phone'] ?? '',
        licenseNumber: json['licenseNumber'],
        licenseCategory: json['licenseCategory'],
        licenseExpiresAt:
            json['licenseExpiresAt'] != null ? DateTime.tryParse(json['licenseExpiresAt']) : null,
        status: json['status'] ?? 'pending',
      );
}

/// True when [date] is within [days] days from now (or already past).
bool isExpiringSoon(DateTime? date, {int days = 30}) {
  if (date == null) return false;
  final diff = date.difference(DateTime.now()).inDays;
  return diff <= days;
}

bool isExpired(DateTime? date) {
  if (date == null) return false;
  return date.isBefore(DateTime.now());
}
